src/btree/bt_split.c

```diff
@@ -986,6 +985,18 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
 					    session, 0, ikey, size));
 					WT_MEMSIZE_ADD(parent_decr, size);
 				}
+				/*
+				 * The page_del structure can be freed
+				 * immediately: it is only read when the ref
+				 * state is WT_REF_DELETED.  The size of the
+				 * structures wasn't added to the parent: don't
+				 * decrement.
+				 */
+				if (next_ref->page_del != NULL) {
+					__wt_free(session,
+					    next_ref->page_del->update_list);
+					__wt_free(session, next_ref->page_del);
+				}
```

Comment:
@michaelcahill, I think this code should appear at line 893 where we flip the WT_REF state, we never reset the state on failure, so we should clean up the memory at the same time. Something like this:
```
diff --git a/src/btree/bt_split.c b/src/btree/bt_split.c
index 018136d..e64f2d5 100644
--- a/src/btree/bt_split.c
+++ b/src/btree/bt_split.c
@@ -893,6 +893,19 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
                    WT_ATOMIC_CAS4(next_ref->state,
                    WT_REF_DELETED, WT_REF_SPLIT))
                        deleted_entries++;
+
+                       /*
+                        * The page_del structure is freed immediately: it is
+                        * only read when the ref state is WT_REF_DELETED and
+                        * we do not reset that value on failure. The size of
+                        * the structures wasn't added to the parent: don't
+                        * decrement.
+                        */
+                       if (next_ref->page_del != NULL) {
+                               __wt_free(session,
+                                   next_ref->page_del->update_list);
+                               __wt_free(session, next_ref->page_del);
+                       }
        }
 
        /*
@@ -985,18 +998,6 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
                                            session, 0, ikey, size));
                                        WT_MEMSIZE_ADD(parent_decr, size);
                                }
-                               /*
-                                * The page_del structure can be freed
-                                * immediately: it is only read when the ref
-                                * state is WT_REF_DELETED.  The size of the
-                                * structures wasn't added to the parent: don't
-                                * decrement.
-                                */
-                               if (next_ref->page_del != NULL) {
-                                       __wt_free(session,
-                                           next_ref->page_del->update_list);
-                                       __wt_free(session, next_ref->page_del);
-                               }
                        }
 
                        WT_TRET(__split_safe_free(
```